src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java

```diff
@@ -25,147 +25,108 @@
+    private static void sendNotifications(Run<?, ?> build, TaskListener listener)
             throws IOException, InterruptedException {
-        createBuildCommitStatus(build, listener, bitbucket);
-    }
-
-    private static BitbucketNotifier getNotifier(BitbucketApi bitbucket) {
-        return new BitbucketChangesetCommentNotifier(bitbucket);
-    }
+        final SCMSource s = SCMSource.SourceByItem.findSource(build.getParent());
+        if (!(s instanceof BitbucketSCMSource)) {
+            return;
+        }
+        BitbucketSCMSource source = (BitbucketSCMSource) s;
+        if (new BitbucketSCMSourceContext(null, SCMHeadObserver.none())
+                .withTraits(source.getTraits())
+                .notificationsDisabled()) {
+            return;
+        }
+        SCMRevision r = SCMRevisionAction.getRevision(build);  // TODO JENKINS-44648 getRevision(s, build)
+        String hash = getHash(r);
+        if (hash == null) {
+            return;
+        }
+        if (r instanceof PullRequestSCMRevision) {
+            listener.getLogger().println("[Bitbucket] Notifying pull request build result");
+            createStatus(build, listener, source.buildBitbucketClient((PullRequestSCMHead) r.getHead()), hash);
 
-    @CheckForNull
-    private static BitbucketSCMSource lookUpSCMSource(Run<?, ?> build) {
-        ItemGroup<?> multiBranchProject = build.getParent().getParent();
-        if (multiBranchProject instanceof SCMSourceOwner) {
-            SCMSourceOwner scmSourceOwner = (SCMSourceOwner) multiBranchProject;
-            BitbucketSCMSource source = lookUpBitbucketSCMSource(scmSourceOwner);
-            if (source != null) {
-                return source;
-            }
+        } else {
+            listener.getLogger().println("[Bitbucket] Notifying commit build result");
+            createStatus(build, listener, source.buildBitbucketClient(), hash);
         }
-        return null;
     }
 
-
-
-    /**
-     * It is possible having more than one SCMSource in our MultiBranch project.
-     * TODO: Does it make sense having more than one of the same type?
-     *
-     * @param scmSourceOwner An {@link Item} that owns {@link SCMSource} instances.
-     * @return A source or null
-     */
     @CheckForNull
-    private static BitbucketSCMSource lookUpBitbucketSCMSource(final SCMSourceOwner scmSourceOwner) {
-        for (SCMSource scmSource : scmSourceOwner.getSCMSources()) {
-            if (scmSource instanceof BitbucketSCMSource) {
-                return (BitbucketSCMSource) scmSource;
-            }
+    private static String getHash(@CheckForNull SCMRevision revision) {
+        if (revision instanceof PullRequestSCMRevision) {
+            // unwrap
+            revision = ((PullRequestSCMRevision) revision).getPull();
         }
-        return null;
-    }
-    
-    private static void sendNotifications(Run<?, ?> build, TaskListener listener)
```

Comment:
Part of the issue here is that it is actually also fixing bugs that I found during the refactoring... and then after I found them I noticed that others had raise JIRA issues for these bugs... and in some cases also PRs... e.g. https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/52 was a parallel fix to the fix I had developed here